commit cd5bb7c87b1891e21c095b3ecd549ef44a067c08
Author: Andreas Dilger <adilger@whamcloud.com>
Date:   Thu May 31 16:39:54 2012 -0400

    e2fsck: allow checking on mounted root fs
    
    Commit 732e26b98e5c79a4298dbe341f43b54b354bb241 added checks to
    prevent e2fsck from being run in filesystem-modifying mode against
    a mounted or otherwise busy device, due to several bug reports of
    users doing this even with the verbose warnings in check_mount().
    
    However, it also prevented e2fsck from checking a mounted root
    filesystem, which will prevent the node from booting.  Once again
    allow e2fsck to run against the mounted root filesystem if it is
    also mounted read-only at the time.
    
    Signed-off-by: Andreas Dilger <adilger@whamcloud.com>
    Signed-off-by: Theodore Ts'o <tytso@mit.edu>

Index: e2fsprogs-1.41.12/e2fsck/unix.c
===================================================================
--- e2fsprogs-1.41.12.orig/e2fsck/unix.c
+++ e2fsprogs-1.41.12/e2fsck/unix.c
@@ -1039,8 +1039,12 @@ restart:
 #endif
 		io_ptr = unix_io_manager;
 	flags = EXT2_FLAG_NOFREE_ON_ERROR;
-	if ((ctx->options & E2F_OPT_READONLY) == 0)
-		flags |= EXT2_FLAG_RW | EXT2_FLAG_EXCLUSIVE;
+	if ((ctx->options & E2F_OPT_READONLY) == 0) {
+		flags |= EXT2_FLAG_RW;
+		if (!(ctx->mount_flags & EXT2_MF_ISROOT &&
+		      ctx->mount_flags & EXT2_MF_READONLY))
+			flags |= EXT2_FLAG_EXCLUSIVE;
+	}
 
 	retval = try_open_fs(ctx, flags, io_ptr, &fs);
 
